API Review Gate blocks every open PR whose merge-base predates cmd/api-gate

The `API Review Gate` required check (`.github/workflows/api-review-gate.yaml`) fails on every open PR whose merge-base with `main` predates the gate's own landing commit (`fa871249c`, 2026-07-01), and it fails before evaluating any API content. This currently red-blocks the entire pre-gate open-PR backlog.

## Mechanism

The "Materialize base checkout at the merge base" step builds the detector from `git merge-base(head, origin/main)` rather than the `main` tip — intentional, to avoid base-drift inventing or hiding API findings. The "Build api-gate" step then does:

```sh
if [ -d "${RUNNER_TEMP}/base/cmd/api-gate" ]; then
  build_dir="${RUNNER_TEMP}/base"          # trusted base
else
  build_dir="${GITHUB_WORKSPACE}"          # head fallback (bootstrap)
fi
( cd "${build_dir}" && go build -o "${RUNNER_TEMP}/api-gate" ./cmd/api-gate )
```

For any PR whose merge-base is older than `fa871249c`, `cmd/api-gate` exists in neither the merge-base checkout nor the (equally old) PR head, so the build fails:

```text
##[warning]api-gate not present on the base branch yet; building from head (bootstrap)...
stat .../cmd/api-gate: directory not found
##[error]Process completed with exit code 1
```

The check reports FAILURE before any API detection runs.

## Evidence

- A non-API PR that only adds a container `resources` block (#2614) fails the gate with every other check green — so it is not gating on API content, it is failing to build.
- PRs branched after 2026-07-01 (e.g. #3184, #3182, #3178, #3177) pass the gate.
- The gate re-triggers on `pull_request_review`, so an approving review re-runs it and it re-fails from the same old merge-base — neither a review nor a plain re-run clears it.

## Impact

Every open PR created before 2026-07-01 is blocked on this required check regardless of what it changes. The only per-PR workaround is a rebase onto current `main`, which moves the merge-base past `fa871249c` — a lot of manual toil across the whole backlog.

## Suggested fix

When `cmd/api-gate` is absent from the merge-base checkout, fall back to building the detector from the `main` tip (or a pinned trusted ref) instead of the PR head. The head fallback only helps during the bootstrap window before the tool exists anywhere, which has already passed, so it can no longer succeed. Building from `main` tip when the merge-base lacks the tool would unblock the backlog without requiring every open PR to rebase, while keeping the "analyze head as data, not code" property. This is distinct from #3184, which addresses detection false-negatives rather than this build failure.

--- a/.github/workflows/api-review-gate.yaml
+++ b/.github/workflows/api-review-gate.yaml
@@ -69,16 +69,30 @@ jobs:
       - name: Build api-gate
         run: |
           set -euo pipefail
-          # Prefer building the detector from the trusted base checkout so a PR
-          # cannot neuter the gate by editing its own detector code — the tool
-          # analyzes the head as *data*, not as code. Fall back to the head
-          # checkout only while the tool has not yet landed on the base branch
-          # (bootstrap); in that window CODEOWNERS on cmd/api-gate,
-          # internal/apigate, and this workflow keeps tampering behind an
-          # API-owner review.
+          # Build the detector from a trusted checkout so a PR cannot neuter the
+          # gate by editing its own detector code — the tool analyzes the head as
+          # *data*, not as code. This selection stays inline in the workflow (an
+          # API-owner-protected file that is always present) rather than in a
+          # helper script: a helper would be read from the untrusted PR head,
+          # which both breaks PRs whose head predates the helper and lets a PR
+          # steer the selection.
+          #
+          # Precedence: prefer the merge-base checkout; when the merge base
+          # predates the commit that introduced the gate (cmd/api-gate absent
+          # there and in the equally old head) fall back to the base branch tip,
+          # which is equally trusted (protected branch, not PR-authored) and
+          # current — this unblocks PRs whose merge base is older than the gate
+          # without forcing a rebase. Bootstrap from head only while the tool
+          # exists nowhere on the base branch yet; in that window CODEOWNERS on
+          # cmd/api-gate, internal/apigate, and this workflow keeps tampering
+          # behind an API-owner review.
           if [ -d "${RUNNER_TEMP}/base/cmd/api-gate" ]; then
             build_dir="${RUNNER_TEMP}/base"
-            echo "Building api-gate from the trusted base checkout."
+            echo "Building api-gate from the trusted merge-base checkout."
+          elif git cat-file -e "origin/${BASE_REF}:cmd/api-gate" 2>/dev/null; then
+            git worktree add --detach "${RUNNER_TEMP}/base-tip" "origin/${BASE_REF}"
+            build_dir="${RUNNER_TEMP}/base-tip"
+            echo "Building api-gate from the ${BASE_REF} tip; the merge base predates the gate."
           else
             build_dir="${GITHUB_WORKSPACE}"
             echo "::warning::api-gate not present on the base branch yet; building from head (bootstrap). CODEOWNERS enforces API-owner review on the gate's own files."

